tests/bsim/bluetooth/host/gatt/device_name/src/server.c

```diff
@@ -0,0 +1,147 @@
+
+static DEFINE_FLAG(security_changed_flag);
+
+static struct bt_conn_cb server_conn_cb;
+
+static void security_changed(struct bt_conn *conn, bt_security_t level, enum bt_security_err err)
+{
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr_str, sizeof(addr_str));
+
+	TEST_ASSERT(err == 0, "Security update failed: %s level %u err %d", addr_str, level, err);
+
+	LOG_DBG("Security changed: %s level %u", addr_str, level);
+	SET_FLAG(security_changed_flag);
+}
+
+static void init_server_conn_callbacks(void)
+{
+	int err;
+
+	server_conn_cb.connected = NULL;
+	server_conn_cb.disconnected = NULL;
+	server_conn_cb.security_changed = security_changed;
+	server_conn_cb.identity_resolved = NULL;
+
+	err = bt_conn_cb_register(&server_conn_cb);
+	TEST_ASSERT(err == 0, "Failed to set server conn callbacks (err %d)", err);
+}
+
+static void connect_and_set_security(struct bt_conn **conn)
+{
+	int err;
+	bt_addr_le_t client = {};
+
+	err = bt_testlib_scan_find_name(&client, ADVERTISER_NAME);
+	TEST_ASSERT(err == 0, "Failed to start scan (err %d)", err);
+
+	err = bt_testlib_connect(&client, conn);
+	TEST_ASSERT(err == 0, "Failed to initiate connection (err %d)", err);
+
+	err = bt_conn_set_security(*conn, BT_SECURITY_L2);
+	TEST_ASSERT(err == 0, "Failed to set security (err %d)", err);
+
+	WAIT_FOR_FLAG(security_changed_flag);
+}
+
+void server_procedure(void)
+{
+	/* Test purpose:
+	 *
+	 * Verifies that writing to the GAP Device Name characteristic correctly
+	 * update the device name.
+	 *
+	 * Two devices:
+	 * - `server`: GATT server, connect and elevate security
+	 * - `client`: GATT client, when connected will look for the GAP Device
+	 *   Name characteristic handle and then will send a GATT write with a
+	 *   new name
+	 *
+	 * [verdict]
+	 * - the server device name has been updated by the client
+	 */
+
+	int err;
+	struct bt_conn *conn = NULL;
+
+	bool names_are_matching;
+
+	uint8_t expected_name[CONFIG_BT_DEVICE_NAME_MAX];
+
+	/* add one for the null character */
+	char original_name[CONFIG_BT_DEVICE_NAME_MAX + 1];
+	char new_name[CONFIG_BT_DEVICE_NAME_MAX + 1];
+
+	const char *name;
+
+	generate_name(expected_name, CONFIG_BT_DEVICE_NAME_MAX);
+
+	TEST_START("server");
+
+	bk_sync_init();
+
+	err = bt_enable(NULL);
+	TEST_ASSERT(err == 0, "Cannot enable Bluetooth (err %d)", err);
+
+	LOG_DBG("Bluetooth initialized");
+
+	err = bt_set_name("Server Super Name");
+	TEST_ASSERT(err == 0, "Failed to set the name (err %d)", err);
+
+	name = bt_get_name();
+	memcpy(original_name, name, strlen(name) + 1);
+
+	init_server_conn_callbacks();
+
+	connect_and_set_security(&conn);
+
+	/* wait for client to do gatt write */
+	bk_sync_wait();
+
+	name = bt_get_name();
+	memcpy(new_name, name, strlen(name) + 1);
+
+	LOG_DBG("Original Device Name: %s", original_name);
+	LOG_DBG("New Device Name: %s", new_name);
+
+	names_are_matching =
+		memeq(expected_name, sizeof(expected_name), new_name, strlen(new_name));
```

Comment:
Creating https://github.com/zephyrproject-rtos/zephyr/pull/84159 only to provide for a single test here, is overkill.

Reply:
![image](https://github.com/user-attachments/assets/eab23953-b341-4a33-b5c6-896afdd5fdbb)

As pointed by @Thalley `memcmp.* == 0` is commonly used and according to him 10% of those usage check the size before. So I would say that there is a use case.

Reply:
BTW, `str(n)cmp` can be used here.

```suggestion
	names_are_matching = strncmp(expected_name, new_name, CONFIG_BT_DEVICE_NAME_MAX);
```

Reply:
So if it should be tested that is can't be greater than `CONFIG_BT_DEVICE_NAME_MAX` you already have a bigger problem with using `strlen`, as this can go out-of-bounds.

```c
	TEST_ASSERT(new_name[CONFIG_BT_DEVICE_NAME_MAX] == '\0', "new_name not NULL-terminated");
	names_are_matching = strncmp(expected_name, new_name, CONFIG_BT_DEVICE_NAME_MAX);
```

Reply:
That's more complicated. `memeq` is ensuring that the two lengths are equal.

Can we just have `memeq`, it's easier to use, makes the code shorter, makes it safer by avoiding reading in out-of-bound memory.